### PR TITLE
[Kernel] Slicing Batched Graphs

### PR DESCRIPTION
**Pull request by mufeili**
## Description
This PR allows retrieving individual component graphs from a batched graph, motivated by #2851 .

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: include/dgl/aten/coo.h [hunk 1]
```diff
@@ -545,6 +545,47 @@ std::vector<COOMatrix> DisjointPartitionCooBySizes(
...
+ *
+ * ret = [[0, 0],
+ *        [1, 0],
+ *        [0, 1]]
+ * COOMatrix_ret.num_rows : 3
+ * COOMatrix_ret.num_cols : 2
+ *
+ * \param coo COOMatrix to slice.
+ * \param edge_cumsum Number of edges in the chunk
+ * \param src_vertex_cumsum Number of src vertices in the chunk.
```
[the remark this answers is not in the document]
Comment (mufeili) [reply include/dgl/aten/coo.h:579]: Yes, do you prefer renaming `edge_cumsum`, `src_vertex_cumsum`, and `dst_vertex_cumsum` to `edge_range`, `src_vertex_range`, and `dst_vertex_range`?

File: python/dgl/batch.py [hunk 2]
```diff
@@ -416,6 +416,61 @@ def unbatch(g, node_split=None, edge_split=None):
...
+        else:
+            start_eid.append(F.as_scalar(F.sum(F.slice_axis(batch_num_edges, 0, 0, gid), 0)))
+
+    # Slice graph structure
+    gidx = slice_gidx(g._graph, utils.toindex(num_nodes), utils.toindex(start_nid),
+                      utils.toindex(num_edges), utils.toindex(start_eid))
+    retg = DGLHeteroGraph(gidx, g.ntypes, g.etypes)
+
+    # Slice node features
+    for ntid, ntype in enumerate(g.ntypes):
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/batch.py:460]: How about adding an optional flag?

File: src/array/union_partition.cc [hunk 3]
```diff
@@ -222,5 +259,42 @@ std::vector<CSRMatrix> DisjointPartitionCsrBySizes(
   return ret;
 }
 
+CSRMatrix CSRSliceContiguousChunk(
+  const CSRMatrix &csr,
+  const std::vector<uint64_t> &edge_cumsum,
+  const std::vector<uint64_t> &src_vertex_cumsum,
+  const std::vector<uint64_t> &dst_vertex_cumsum) {
+  int64_t indptr_len = src_vertex_cumsum[1] - src_vertex_cumsum[0] + 1;
+  IdArray result_indptr = IdArray::Empty({indptr_len}, csr.indptr->dtype, csr.indptr->ctx);
```
Comment (classicsong) [inline src/array/union_partition.cc:268]: Check if the result_indptr is all zero.

Comment (mufeili) [reply src/array/union_partition.cc]: You are right. This is fixed now.

**Review by classicsong — APPROVED** [review]
LGTM

**Pull request merged**